### DISCUSSION TITLE
Charmil Roadmap

### DISCUSSION THREAD

**Discussion by ankithans** [aerogear/charmil#33] [General]
We need to discuss, what will be the SDK doing exactly and list of features, to proceed further.

**Comment by wtrocki** [comment]
Sure. It is better to just have some base code first to leave groundwork for the discussion. 
Alternatively, we can draft some ideas here. 
I think we touchdown on how SDK will work but my take is that we simply need to define core feature. 
Krew is actually complex system if you think about it and with your level of golang it will be hard to move to that so fast. 

My suggestion is to start simple. Actually very simple and write CLI host and for the moment try to embed other commands (for example embed other CLI) - that would be possible without charmil right?

Good example would be to look at the POC that one of our colleagues wrote:
https://github.com/craicoverflow/socket

We really need an extensibility mechanism as soon as possible so starting with something that works is great.

If we ignore problem of downloading command line tool and indexing it and assume that we simply want to include modules into our binary then this might be much easier to do.

**Comment by ankithans** [comment]
> My suggestion is to start simple. Actually very simple and write CLI host and for the moment try to embed other commands (for example embed other CLI) - that would be possible without charmil right?

okay got your point. Will start working on it today 👍

**Comment by ankithans** [comment]
I have tried and embedding works, I created sample CLI's, installed binary and combined them (played a lot with them).
With the help of above POC.

**Comment by wtrocki** [comment]
We can provide more opiniated layer for both plugin and host of cobra commands. Making it somewhat high level version of cobra that resolves extra concerns. I think this issue is about listing some features. My take is that those are less important right now but worth to put some of them to give us glimpse what value this can give:

Example of functionalities for both Host and Plugins:
- Standard for arguments passed to command
- helpers to buld set of CRUD commands without boilerplate
- help topics standarization
- Command examples standarization and validation
- Configuration storage and abstraction for plugin.
- Storage for Secrets (like oauth token)
- Logging and printing messages unified way
- Error handling and validation logic
- Internationalization
- Security (Authorization)
- Feature flags to enable specific plugins
- OpenAPI file mapping to SDKs..
- Dozens more in future if needed.
- Ability to use plugins without compiling them into root CLI (AKA registry)

There could be also plugins specific features:
- Dependency on other plugins and reusing of code
- Provining its own implementations for concerns mentioned above

As we see here scope for this work can be big so my take is to not think about features - like building car or rocket we might start with engine and frame to see how we will be able to fit those in the next phase.

All of that can be some extra capabilities that will work on top the small engine/core lib.
Making core lib that does plugin piping is essential to start adding any feature.

## Phase 1 Rocket Science of Core SDK

I think we can take two approaches. Each of those might require separate discussion:

1) Allow plugins to implement cobra structures and somewhat validate it on root.
2) Provide wrapper structure without any strong dependency on Cobra being even present in the plugin and then in core based of those structures build cobra command.

Nr1 seems like no fuss but it is worth to have open discussion about pros and cons.

## Phase 2 Honest and Hard Work (on features)

Pick and implement some specific capabilities.
Discuss what we should do first. Engage other mentors for learning and validation.
Document every feature. Provide examples etc.

## Phase 3 Integrate and challenge

Try to validate POC with external adopters.
For example we can ask Lead of Managed Services CLI if we could integrate this with RHOAS CLI. See if there is value in integration etc.

If we bring value === Success. If not we can pivot to real problem etc.

## Phase 4 Docs and Promotion

Blogs, Documentation etc. Time for swag.

## Phase 5 Rockstar Open Source Engineer is born

Kinda self explanatory but have picture for visualization:

![image](https://user-images.githubusercontent.com/981838/119998759-ec138a00-bfc8-11eb-8e3e-58919e235b68.png)


This is rough plan that can be validated and adjusted once we reach milestones

**Comment by ankithans** [comment]
Wow! that's just the great insight into the work to be done and the project lifecycle in general. Previously I was confused, where to start and where will we go. Now there is much clarity here. 
So we are at phase1
> Provide wrapper structure without any strong dependency on Cobra being even present in the plugin and then in core based of those structures build cobra command.

looks like we can cover a huge market here! But this is the initial thought, so research is needed.

**Comment by namit-chandwani** [comment]
@wtrocki Very well summarized 👏
I agree with Ankit, especially on the clarity front. This appropriately depicts where we are today and how much we have to accomplish over this summer.

We'll start working now while keeping this roadmap in mind.

**Comment by wtrocki** [comment]
So we have two spikes for doing core sdk - no coincidence here ;) Pick one and create issue for deep dive discussion and implementation.

> Allow plugins to implement cobra structures

@ankithans have been exploring this one already so if you still want it then @namit-chandwani  can take wrapper approach.

**Comment by namit-chandwani** [comment]
> @ankithans have been exploring this one already so if you still want it then @namit-chandwani  can take wrapper approach.
> 
> 

I talked to @ankithans, he wants to go with the 2nd approach (wrapper). So I'll take the first one and create an issue on it as mentioned.

**Comment by wtrocki** [comment]
Great job! Lets move all discussions to other issues so we will keep this one dedicated to roadmap.

[Offtopic]
@namit-chandwani in AeroGear and GSOC  we tend to keep all conversations in the open channels. Lets use Charmil discord even for more casual discussions. I understand that Github issues might be to noisy. Lets avoid PM's etc.

**Comment by namit-chandwani** [comment]
> [Offtopic]
> @namit-chandwani in AeroGear and GSOC  we tend to keep all conversations in the open channels. Lets use Charmil discord even for more casual discussions. I understand that Github issues might be to noisy. Lets avoid PM's etc.

Yes, that was wrong on our part. Won't happen again for sure. 
Will use the Charmil discord channel for all our future discussions :)

**Comment by wtrocki** [comment]
CC @craicoverflow
